File: hnetwork_test.py
import hnetwork as hn;
import decentral_search as ds;
import deg_dist as dd;
import random;
import time;

def test_search(S):
    num_searches = 5000;
    total_cost = 0.0;
    num_successes = 0.0;
    p = 0.25;
    
    start = time.clock();
    for i in xrange(num_searches):
        s = random.choice(N.G.nodes());
        t = random.choice(N.G.nodes());
        cost = ds.search(S, N, s, t, p);
        if cost != float("inf"):
            total_cost = total_cost + cost;
            num_successes = num_successes + 1.0;

    diff = time.clock() - start;
    avg_length = total_cost / num_successes if num_successes > 0 else \
                float("inf");
    
    print "\tsuccess rate =", num_successes / num_searches;
    print "\taverage search length =", avg_length;
    print 
    print "Took", diff, "seconds for", num_searches, "searches";

random.seed();

start = time.clock();
N = hn.HNetwork(2, 5, 2, 100, 1, dd.ConstantDegDist(300));
diff = time.clock() - start;

# print "Nodes =", N.G.nodes(data = True);
# print "Edges =", N.G.edges();
# print

print "Random HNetwork generated:";
print "\tH =", N.H;
print "\tl =", N.l;
print "\tb =", N.b;
print "\tg =", N.g;
print "\ta =", N.a;
print "\tnumber_of_nodes =", N.G.number_of_nodes();
print "\tnumber_of_edges =", N.G.number_of_edges();
print
print "Took", diff, "seconds";

print
print "Perfect Decentralized Search:";
test_search(ds.PerfectDecentralizedSearch());

alpha = 100000;
print
print "Imperfect Decentralized Search:";
print "\talpha =", alpha;
test_search(ds.SoftmaxDecentralizedSearch(alpha));

File: cpp/hnetwork_test.cpp
#include <iostream>
#include <ctime>
#include <cstdlib>

#include "deg_dist.h"
#include "hnetwork.h"
#include "search.h"

using namespace std;

#define NUM_SEARCHES 100000
#define P 0.25

static void test_search(DecentralizedSearch& S, HNetwork& N);

int main() {
  srand(time(NULL));

  clock_t start = clock();
  ConstantDegDist dd(99);
  // PowerLawDegDist dd(2.5, 40, 1500);
  HNetwork hn(2, 13, 2, 100, 1, dd);
  double diff = (1.0*clock() - start) / CLOCKS_PER_SEC;

  cout << "Random HNetwork generated:" << endl;
  cout << "\tH = " << hn.getH() << endl;
  cout << "\tl = " << hn.getl() << endl;
  cout << "\tb = " << hn.getb() << endl;
  cout << "\tg = " << hn.getg() << endl;
  cout << "\ta = " << hn.geta() << endl;
  cout << "\tnumber_of_nodes = " << hn.getN() << endl;
  cout << "\tnumber_of_edges = " << hn.getE() << endl;
  cout << endl << "Took " << diff << " seconds" << endl;

  cout << endl << "Perfect Decentralized Search:" << endl;
  PerfectDecentralizedSearch p;
  test_search(p, hn);

  /* double beta = 6;
  cout << endl << "Imperfect Decentralized Search (beta = "
       << beta << "):" << endl;
  SoftmaxDecentralizedSearch d(beta);
  test_search(d, hn); */

  return 0;
}

static void test_search(DecentralizedSearch& S, HNetwork& N) {
  double total_cost = 0.0;
  int num_successes = 0;

  clock_t start = clock();
  int i;
  for (i = 0; i < NUM_SEARCHES; i++) {
    int s = rand() % N.getN();
    int t = rand() % N.getN();
    int cost = S.search(N, s, t, P);
    if (cost != INF) {
      total_cost += cost;
      num_successes++;
    }
  }
  double diff = (1.0*clock() - start) / CLOCKS_PER_SEC;
  double avg_length = INF;
  if (num_successes > 0) {
    avg_length = total_cost / num_successes;
  }

  cout << "\tsuccess rate = " << 1.0*num_successes / NUM_SEARCHES << endl;
  cout << "\taverage search length = " << avg_length << endl;
  cout << "\tTook " << diff << " seconds for "
       << NUM_SEARCHES << " searches" << endl;
}
